### ISSUE TITLE
Chapter 2 | How To ? | Getting Pandas Object from Sklearn Pipeline Output

### ISSUE THREAD

**Issue by raamav** [ageron/handson-ml#411]
Hi @ageron :
Thanks a ton for putting together an excellent book, I am learning a lot and its been very satisfying.

I am interested in analyzing the Pipeline output, however, as you know the output of a pipeline is a numpy array which loses the information pertaining to the columns.

Another related goal is to analyze the model output in greater detail.

I see that you have provided a code where we manually construct the dataframe from the numPy array.

`housing_extra_attribs = pd.DataFrame(
    housing_extra_attribs,
    columns=list(housing.columns)+["rooms_per_household", "population_per_household"])
housing_extra_attribs.head()`

I wanted to get your thoughts on the above. 

To reiterate, the goals are:

1. To get the pipeline output as a data frame in a more 'automated' fashion
2. To be able to interpret the linear regression results in a comprehensive manner (e.g. r-squared, the significance of parameters, relative weight of parameters )

Thanks for your help
Rama

PS::

I have a question re. Pipelines. Is building pipelines needed when the model is built from an exploratory/learning POV? (i.e. the model is not to be deployed in a production environment)

**Comment by raamav** [comment]
Hi @ageron 
I figured out a nifty piece of code that creates a data frame with the sorted columns

`coefficients = pd.concat([pd.DataFrame(all_attribs),pd.DataFrame(np.transpose(lin_reg.coef_))], axis = 1)
coefficients.columns = ["attribute","coefficients"]

coefficients = coefficients.sort_values(['coefficients'], ascending = [False])`

I'll await your response to know if you have further comments, else I'll close out the issue.

Thanks again.

**Comment by ageron** [comment]
Hi @raamav ,
Thanks for your kind words, I'm glad you find the book useful!
Unfortunately, Scikit-Learn was created before Pandas existed, which explains why it offers little support for it. So when you pass a DataFrame to a Scikit-Learn model, you often get a NumPy array out, thus without column information, as you noted. There may be some tools to help preserve the column names, but I have not used them. When I want to recover the column names, I do it manually, as shown in the code example you pointed to.
Regarding your second question, in a LinearRegression model, the weights are listed in the same order as the input columns. For example:

```pycon
>>> from sklearn.linear_model import LinearRegression
>>> import pandas as pd
>>> import numpy as np
>>> X_train = pd.DataFrame({"age": [44., 23., 35.], "weight": [80., 77., 68.]})
>>> y_train = np.array([20., 19., 15.])
>>> lin_reg = LinearRegression()
>>> lin_reg.fit(X_train, y_train)
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)
>>> lin_reg.score(X_train, y_train)
1.0
>>> coefs = pd.DataFrame(lin_reg.coef_.reshape(1, -1), columns=X_train.columns)
>>> coefs
        age    weight
0 -0.013333  0.426667
```

Note that the `score()` method computes the R2 score (see [doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.score)).

Regarding your last question (in the PS), you never _have_ to use pipelines. They're just convenient ways of, well, pipelining multiple preprocessing steps. I think it's good practice, but you're welcome to use any other preprocessing system you prefer, Scikit-Learn is very modular, you can just use any piece you like without using the rest.

Hope this helps!

**Comment by raamav** [comment]
Thanks a ton @ageron
